compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -907,18 +907,34 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             };
             let ty = self.normalize(expr.span, ty);
             if self.can_coerce(found, ty) {
+                let span = self.find_suggestable_ancestor_span(expr);
+
                 err.multipart_suggestion(
                     "you might have meant to return this value",
                     vec![
-                        (expr.span.shrink_to_lo(), "return ".to_string()),
-                        (expr.span.shrink_to_hi(), ";".to_string()),
+                        (span.shrink_to_lo(), "return ".to_string()),
+                        (span.shrink_to_hi(), ";".to_string()),
                     ],
                     Applicability::MaybeIncorrect,
                 );
             }
         }
     }
 
+    /// Given an expression (possibly as a product of multiple macro expansions in a chain), find
+    /// the closest ancestor span which is no longer the source of a macro expansion.
+    fn find_suggestable_ancestor_span(&self, expr: &Expr<'_>) -> Span {
+        let mut node_id = self.tcx.hir().parent_id(expr.hir_id);
+        let mut parent_id_iter = self.tcx.hir().parent_iter(node_id);
+        while let Some((id, _)) = parent_id_iter.next() && self.tcx.hir().span(id).from_expansion() {
+            node_id = id;
+        }
+        let parent = self.tcx.hir().get_enclosing_scope(node_id).unwrap_or(node_id);
+        let parent_span = self.tcx.hir().span(parent);
+        let span = expr.span.find_ancestor_inside(parent_span).unwrap_or(expr.span);
+        span
+    }
```

Comment:
What happens if the function itself is generated by a macro?
What happens if a local macro generates non-trivial code? Should we suggest inside that macro?
Should we stop the parent walk when we stop seeing expressions?

To be clear: I think this heuristic is too strong, and should be refined a bit.

Reply:
> What happens if the function itself is generated by a macro?

For something like

```rust
macro_rules! entire_fn {
    () => {
        pub fn baz(x: &str) -> Result<(), Box<dyn std::error::Error>> {
            Err(format!("error: {x}"))
            //~^ ERROR mismatched types
        }
    }
}

entire_fn!();
```

I could get the span to be inside `entire_fn` via `find_ancestor_in_same_ctxt` instead of `find_ancestor_inside`.

> What happens if a local macro generates non-trivial code? Should we suggest inside that macro?

I'm having a hard time trying to come up with a test example for this, do you have an example in mind?

> Should we stop the parent walk when we stop seeing expressions?

I'm not quite sure about this... I don't think it would be right if we stopped the parent walk when we stop seeing expressions? The parent could very well be items and not expressions?

Reply:
> I'm having a hard time trying to come up with a test example for this, do you have an example in mind?

In your test, having the macro expand to `Err(format!(...))`, instead of just the `format!`.

Reply:
Yeah unfortunately I don't think this can work...

Given something like

```rust
macro_rules! nontrivial {
    ($x: expr) => {
        Err(format!("error: {}", $x))
        //~^ ERROR mismatched types
    }
}

pub fn qux(x: &str) -> Result<(), Box<dyn std::error::Error>> {
    nontrivial!(x)
}
```

my current code will suggest `nontrivial!(x).into()` which isn't correct. Should we just suppress `Into::into` suggestion when `expr`'s span comes from a macro expansion...?

Reply:
I think it's okay to continue making a suggestion for the macro code, perhaps only if it's a locally-defined macro (if that's easy to check here). I don't see much value in making that suggestion if the macro is defined in an external crate.